File: src/main/java/com/huffnest/HuffNestFile.java
package com.huffnest;

import com.huffnest.bytetree.ByteTree;
import com.huffnest.bytetree.ByteTreeFactory;
import com.huffnest.bytetree.ByteTreePath;
import com.huffnest.bytetree.ByteTreePath.ByteTreePathIterator;
import com.huffnest.bytetree.ByteTreeSerializer;
import com.huffnest.io.BitReader;
import com.huffnest.io.BitWriter;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.file.Path;

public class HuffNestFile {

  public HuffNestFile(Path inputFilePath, Path outputFilePath) {
    this.inputFilePath = inputFilePath;
    this.outputFilePath = outputFilePath;
  }

  private Path inputFilePath;
  private Path outputFilePath;

  public void compress(int iterationCount) throws IOException {
    int iteration = 1;

    BitReader br = new BitReader(inputFilePath);

    ByteTreeFactory.NewByteTreeBuilder builder = ByteTreeFactory.New();

    System.out.println("Building tree...");

    while (br.hasNextByte()) {
      builder.pushByte(br.nextByte());
    }

    ByteTree tree = builder.build();

    System.out.println("Tree Built!");

    BitWriter bw = new BitWriter(outputFilePath);

    System.out.println("Serializing Tree...");

    byte[] serializedTree = new ByteTreeSerializer().serialize(tree);

    System.out.println("Tree Serialized!");

    bw.pushInt(serializedTree.length);
    bw.pushInt(iteration);

    for (byte b : serializedTree) {
      bw.pushByte(b);
    }

    //room for future improvements:
    // bw.pushInt(0);
    // bw.pushInt(0);

    br = new BitReader(inputFilePath);
    System.out.println("Compressing file...");

    while (br.hasNextByte()) {
      byte b = br.nextByte();
      ByteTreePathIterator pathIterator = tree.getPathToByte(b).iterator();

      while (pathIterator.hasNext()) {
        bw.pushBit((byte) (pathIterator.nextIsLeft() ? 0 : 1));
      }
    }

    byte pbc = bw.close();

    System.out.println("File compressed!");
  }

  public void decompress() throws IOException {
    BitReader br = new BitReader(inputFilePath);
    BitWriter bw = new BitWriter(outputFilePath);

    int treeSize = br.nextInt();
    int iterationsLeft = br.nextInt();

    byte[] serializedTree = new byte[treeSize];

    for (int i = 0; i < treeSize; i++) {
      serializedTree[i] = br.nextByte();
    }

    ByteTree tree = new ByteTreeSerializer().deserialize(serializedTree);

    System.out.println("Decompressing file...");
    ByteTreePath path = ByteTreePath.empty();
    while (br.hasNextBit()) {
      if (!br.hasNextBit()) break;
      path.pushDirectionBit(br.nextBit());
      if (tree.hasByteAtPath(path)) {
        bw.pushByte(tree.getByteAtPath(path));
        path = ByteTreePath.empty();
      }
    }
    bw.close();

    //clean up messed up file end

    path = ByteTreePath.empty();
    byte paddingByteCount = 0;
    for (int i = 0; i < 6; i++) {
      path.pushDirectionBit((byte) 0);
      if (tree.hasByteAtPath(path)) {
        path = ByteTreePath.empty();
        paddingByteCount++;
      }
    }

    RandomAccessFile raf = new RandomAccessFile(outputFilePath.toFile(), "rw");
    long length = raf.length();

    long newLength = Math.max(0, length - paddingByteCount);
    raf.setLength(newLength);
    raf.close();
  }
}

File: HuffNest/HuffNestFile.cs
using HuffmanTree;
using HuffmanTree.Factory;
using IO;
using Util;

public class HuffNestFile(string inputFilePath, string outputFilePath)
{
    public async Task Compress(int iterationCount)
    {
        int iteration = 1;

        BitReader br = new BitReader(inputFilePath);

        var treeBuilder = TreeFactory.New();

        Console.WriteLine("Building tree...");

        while (br.HasNextByte)
        {
            treeBuilder.PushByte(await br.GetNextByteAsync());
        }
        br.Close();

        Tree tree = treeBuilder.Build();

        byte x = (byte)'1';
        for (int i = 0; i < 6; i++)
        {
            tree.TryGetPathToByte(x, out var pth);
            tree.TryGetByteAtPath(pth, out var b);
            x++;
            Console.WriteLine((char)b + "/" + b + " - " + pth);
        }

        Console.WriteLine("Tree Built!");

        BitWriter bw = new BitWriter(outputFilePath);

        Console.WriteLine("Serializing Tree...");

        byte[] serializedTree = TreeSerializer.Serialize(tree);
        Tree tree2 = TreeSerializer.Deserialize(serializedTree);

        Console.WriteLine("Tree Serialized!");

        await bw.WriteIntAsync(serializedTree.Length);
        await bw.WriteIntAsync(iteration);
        await bw.WriteByteAsync(0); //space for byte that describes padding bit count at the end of file

        foreach (var b in serializedTree)
        {
            await bw.WriteByteAsync(b);
        }

        //room for future improvements:
        // bw.pushInt(0);
        // bw.pushInt(0);

        Console.WriteLine("Compressing file...");
        br = new BitReader(inputFilePath);

        while (br.HasNextByte)
        {
            byte b = await br.GetNextByteAsync();
            if (tree.TryGetPathToByte(b, out TreePath path))
            {
                Console.WriteLine("compress " + (char)b + "/" + b + " to " + path);
                foreach (Bit bit in path.GetEnumerator())
                {
                    await bw.WriteBitAsync(bit);
                }
            }
            else
            {
                throw new Exception("failed to get path to byte");
            }
        }
        br.Close();

        byte pbc = await bw.CloseAsync();

        if (pbc > 0)
        {
            using (
                var fs = new FileStream(
                    outputFilePath,
                    FileMode.Open,
                    FileAccess.Write,
                    FileShare.None
                )
            )
            {
                fs.Seek(8, SeekOrigin.Begin);
                fs.WriteByte(pbc);
                fs.Close();
            }
        }

        Console.WriteLine("File compressed!");
    }

    public async Task Decompress()
    {
        BitReader br = new BitReader(inputFilePath);
        BitWriter bw = new BitWriter(outputFilePath);

        int treeSize = await br.NextIntAsync();
        int iterationsLeft = await br.NextIntAsync();

        byte[] serializedTree = new byte[treeSize];

        byte pbc = await br.GetNextByteAsync();

        for (int i = 0; i < treeSize; i++)
        {
            serializedTree[i] = await br.GetNextByteAsync();
        }

        Tree tree = TreeSerializer.Deserialize(serializedTree);

        Console.WriteLine("Decompressing file...");
        UnknownPath path = new();
        while (br.HasNextBit)
        {
            path.PushBit(await br.GetNextBitAsync());
            if (tree.TryGetByteAtPath(path, out byte value))
            {
                await bw.WriteByteAsync(value);
                path.Reset();
            }
        }
        br.Close();
        await bw.CloseAsync();

        //clean up messed up file end

        path.Reset();
        byte paddingByteCount = 0;

        for (int i = 0; i < pbc; i++)
        {
            path.PushBit(new Bit());
            if (tree.TryGetByteAtPath(path, out var _))
            {
                paddingByteCount++;
                path.Reset();
            }
        }

        if (paddingByteCount > 0)
        {
            using (
                var fs = new FileStream(
                    outputFilePath,
                    FileMode.Open,
                    FileAccess.Write,
                    FileShare.None
                )
            )
            {
                long newLength = Math.Max(0, fs.Length - paddingByteCount);
                fs.SetLength(newLength);
                fs.Close();
            }
        }
    }
}
